vissl/losses/barlow_twins_loss.py

```diff
@@ -0,0 +1,124 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+
+import pprint
+
+import torch
+from classy_vision.generic.distributed_util import get_cuda_device_index
+from classy_vision.losses import ClassyLoss, register_loss
+from torch import nn
+from vissl.utils.hydra_config import AttrDict
+
+
+@register_loss("barlow_twins_loss")
+class BarlowTwinsLoss(ClassyLoss):
+    """
+    This is the loss which was proposed in the Barlow Twins https://arxiv.org/abs/2103.03230v1
+    paper.
+    See the paper for the details on the loss.
+
+    Config params:
+        lambda_ (float): weight on the off-diagonal terms. It controls the trade-off between
+            the importance given to the invariance term versus the redundancy reduction term.
+        embedding_dim (int): dimensionality of the representation
+    """
+
+    def __init__(self, loss_config: AttrDict):
+        super(BarlowTwinsLoss, self).__init__()
+
+        self.loss_config = loss_config
+        self.bt_criterion = BarlowTwinsCriterion(
+            self.loss_config.lambda_, self.loss_config.embedding_dim
+        )
+
+    @classmethod
+    def from_config(cls, loss_config: AttrDict):
+        """
+        Instantiates BarlowTwinsLoss from configuration.
+
+        Args:
+            loss_config: configuration for the loss
+
+        Returns:
+            BarlowTwinsLoss instance.
+        """
+        return cls(loss_config)
+
+    def forward(self, output, target):
+        loss = self.bt_criterion(output)
+        return loss
+
+    def __repr__(self):
+        repr_dict = {
+            "name": self._get_name(),
+            "lambda_": self.loss_config.lambda_,
+            "embedding_dim": self.loss_config.embedding_dim
+        }
+        return pprint.pformat(repr_dict, indent=2)
+
+
+class BarlowTwinsCriterion(nn.Module):
```

Comment:
First of all, thank you for the amazing work. All of this looks really really great 👍 

I just wanted to ask an open question more than a review comment (especially since I am not familiar with the initial implementation of @jzbontar) regarding the criterion:

It seems to me that the criterion defined here does not gather the representations across workers (in case of DistributedDataParallel training) which means that each GPU will compute its own cross-correlation matrix. Gathering the representation (as done in the SimCLR criterion) would increase the size of this cross-correlation matrix (by multiplying it with G, G being the number of GPUs used for the training).

So the open question (in particular for @jzbontar) is: should the criterion gather the representation (like SimCLR does for instance?) or is it not part of the Barlow Twins algorithm (and might actually harm the performance, or just increase the memory usage for no effect)?

Reply:
> It seems to me that the criterion defined here does not gather the representations across workers (in case of DistributedDataParallel training) which means that each GPU will compute its own cross correlation matrix, based on its own batch.

Indeed, this is not the case. I totally agree. It wasn't clear to me, so I opted for the naive implementation while waiting for the official implementation to come out.

> or just increase the memory usage for no effect

I am more afraid of the network overhead personally =/. Passing massive [Nx8192] matrices to the master process will have a non negligible effect on training speeds.

BTW, I was wondering if we could do something similar to Batch Norm process groups when computing the SimCLR loss to improve speeds. If we define one process group for each node of a distributed training, we could gather the embedding per process group, compute the loss for each process group and then aggregate the losses over the different groups. If the groups are large enough (meaning each node as multiple GPUs), this should not really impact the performances, shouldn't it? 

For example, if we have a distributed training of octo GPU nodes and each process has a batch size of 512, each process group would apply the SimCLR loss over a local batch size of 8*512 = 4096.

This should lead to speed gains as the large embedding are only shared between GPUs of the same node with very small overhead and we just share a small scalar between the nodes over the network.

Reply:
I am not sure if my rambling was very clear but I just thought about it when reading your comment and wanted your opinion on the subject. :)

Reply:
> Indeed, this is not the case. I totally agree. It wasn't clear to me, so I opted for the naive implementation while waiting for the official implementation to come out.

It was not clear to me either :) Ultimately, I think we should rely on @jzbontar to provide us with the answer.

> I am not sure if my rambling was very clear but I just thought about it when reading your comment and wanted your opinion on the subject. :)

Your comment on SimCLR is quite interesting as well :)

In my understanding, the main goal of augmenting the number of nodes in SimCLR is to augment the total batch size that will go through the SimCLR criterion so that we can reach a big enough batch required to get good representations (in my past experiences, I usually selected the number of nodes such that I get the desired batch size). 

However, there is no real benefit in augmenting the batch size beyond 4096 or 8092 for SimCLR (and it seems to depend on the dataset), so past that point (if we increase the number of nodes further), we could indeed think of aggregating by process groups. This is indeed something that could be chosen by configuration.

Reply:
That's a good point. In our code we synchronize values across all nodes and GPUs. In particular,

1. We use `torch.nn.SyncBatchNorm(..., affine=False)` when normalizing the columns of the embedding matrix
2. We use `torch.distributed.all_reduce(c)` to combine the cross-correlation matrix across all workers

I don't know how omitting these two steps affects the accuracy of Barlow Twins. Nevertheless, I think we should include both steps at first to be consistent with the paper. If it turns out that they are not important, we can remove them and have a per-gpu type loss instead.

Reply:
> We use `torch.distributed.all_reduce(c)` to combine the cross-correlation matrix across all workers

Oh I see! This is way better than gathering the embedding!

> I don't know how omitting these two steps affects the accuracy of Barlow Twins. Nevertheless, I think we should include both steps at first to be consistent with the paper.

Yes, I will patch the implementation when I come back from holidays :)